### ISSUE TITLE
Add shiny to dashboard for interactive data vizualisation

### ISSUE THREAD

**Issue by rempsyc** [rempsyc/busara_dashboard#10]
One reviewer noted,
> A wish list (which may not be technically possible), would be on the home page for you to be able to check the journals you want to include and then have the waffle plot dynamically update.

And another noted,
> I guess the more advanced version of this would be if the website let users pick what variables to plot and let us make new graphs

For interactive data vizualisation, it might be possible to use Shiny in compatibility with `flexdashboard`: https://pkgs.rstudio.com/flexdashboard/articles/shiny.html

**Comment by rempsyc** [comment]
@psforscher at first I got excited when I read that it was possible to embed `shiny` code within `rmarkdown` html files and `flexdashboard`, [e.g.](https://bookdown.org/yihui/rmarkdown/shiny-embedded.html)

> ### Embedded Shiny apps
> Besides embedding individual Shiny inputs and outputs in R Markdown, it is also possible to embed a standalone Shiny application within a document.

[and](https://rstudio.github.io/flexdashboard/articles/shiny.html)

> ### Using shiny with flexdashboard
> By adding Shiny to a flexdashboard, you can create dashboards that enable viewers to change underlying parameters and see the results immediately, or that update themselves incrementally as their underlying data changes (see reactiveFileReader and reactivePoll). This is done by adding runtime: shiny to a standard flexdashboard and then adding one or more input controls and/or reactive expressions that dynamically drive the appearance of the components within the dashboard. Using Shiny with flexdashboard turns a static R Markdown report into an Interactive Document.

However, a bit lower we can also (disappointingly) read:

>  It’s important to note that interactive documents need to be deployed to a Shiny Server to be shared broadly (whereas static R Markdown documents are standalone web pages that can be attached to emails or served from any standard web server).

Which is also confirmed by this [RStudio help forum](https://forum.posit.co/t/embedding-shiny-with-inline-not-rendering-with-html-output/41175/4) where a person had the same issue as we do:

> Let's recall that a shiny document can't be rendered to a static html file like you can in a static website. It is way of creating shiny app from a single Rmd document, but it is still require to be hosted and run when consulted. 

and

> _Note:_ If you are familiar with R Markdown, you might expect RStudio to save an HTML version of an interactive document in your working directory. However, this only works with static HTML documents. Each interactive document must be served by a computer that manages the document. As a result, interactive documents cannot be shared as a standalone HTML file.

So it seems we will have no choice but to rely on a shiny server of some sort.

In the forum they also mention:

> Know that, for including an app into a static document or website, like a bookdown, you can host the shinyapps somewhere (like [shinyapps.io ](http://shinyapps.io/)) and include it in the document with knitr::include_app. see [2.11 Web pages and Shiny apps | bookdown: Authoring Books and Technical Documents with R Markdown](https://bookdown.org/yihui/bookdown/web-pages-and-shiny-apps.html)

So perhaps it is possible to keep the flexdashboard infrastructure, host the shiny app elsewhere, and then embed it inside like described above...

In any case, it IS possible to deploy a shiny app on our own servers, and there are detailed instructions to this effect ([example](https://docs.posit.co/shiny-server/)), but this is too complex even for someone like me. Do you have any specialized IT person at your organization which could help us set this up?

**Comment by psforscher** [comment]
Yes, we do have a software team to work on projects like this. Let me
consult with the software lead to get an estimate of the scope of work.

**Comment by psforscher** [comment]
i did in fact approach one of our developers about deploying the shiny app on one of our servers. here's what he said:

> I will not take a lot of time to learn the deployment -- I've checked it out, the high level deployment steps do not look complicated as long as the programmed dashboard is available.

so i think this is quite possible. perhaps we could think about doing this in the new year? i'd want you to meet with our developer just so that the ask is clear, but it sounds like this could be done in a week or less.

**Comment by rempsyc** [comment]
Alright, is there any way we could do a quick pilot test with a test file before I put time on this? I think the important thing is that I would need to be able to update the dashboard on my week without having to go through IT every time if we are to update the dashboard automatically every week... They said "as long as the programmed dashboard is available". I can make it available on GitHub, but will that work for them?

**Comment by rempsyc** [comment]
@psforscher have a look at this proof of concept: https://rempsyc.shinyapps.io/missing_majority/

**Comment by psforscher** [comment]
wow, this looks great!

i started to fiddle with the options and got a strange error -- an "Ajax error with data.table". i didn't think to take a screenshot and i haven't been able to reproduce this exact error. however, i have been unexpectedly disconnected from the server a few times when clicking around. basically i think this is great as a proof of concept but might require a bit of refinement if we go in this direction.

shall i connect with our developer to see if he can set up the pilot?

**Comment by rempsyc** [comment]
Hi Patrick, thank you! Yes, I was waiting for you to play around with it a bit before giving you the whole info.

First of all, this was a lot of work, but yes, it looks extremely promising! Design-wise, everything seems to be looking similar to the static dashboard, so that's a big win. I was also able to add a persistent side-bar so that you can setup your desired journals and such once and it will apply to all other tabs (this saves _a lot_ of processing time). It is possible to filter by field, journal, and continent vs group (and year range). So this is another big win. A novelty is also now that even the color gauges on the home page react to the journal selection. This is some data visualization we didn't have before!

Some problems:
1. The dashboard takes a lot of time to load in the beginning. This is normal for Shiny apps but slower for this one because of the size of the dashboard (because of our data files).
2. When you first click on the visualization tabs, you get an error at first until the data can finish processing. There is no easy workaround for this as far as I am aware.
3. The free plan of shinyapps.io [is limited to 1 GB of storage space](https://support.posit.co/hc/en-us/articles/219449487--How-much-data-can-I-upload-to-shinyapps-io), meaning we can't upload the general field data from _PloS One_, _Nature_, _Science_, and _PNAS_. So this proof of concept is only for the fields of psychology and economics (the most important anyway for now). Probably if we had our own servers we could get around this limitation (but the dashboard could also be slower to run and process because of this).
4. The free plan also has a limit of 1 GB of RAM, which not only limits the speed of data processing, but will also frequently crash the dashboard when the memory runs over the limit. This frequently leads to the "Disconnected from the server" error when performing costly operations such as as the country scatter plot or the waffle plot with the country flags (it is not impossible to make those work but one must often refresh the page with the correct selection until it works).

![Image](https://github.com/user-attachments/assets/b8a295a8-2f9d-4aca-ba63-ead429bf9178)

Most of these problems could be resolved by upgrading to a paid plan, or better yet, set it up on our own servers.

The good news, and the point of that proof of concept, was to show that with more resources, like on our own servers, things should run quite smoothly ;) (and of course there will always be a bit more refinement to do here and there). And it's also a demonstration for our Busara developer. So yes, please put me in touch!
